src/main/java/com/mongodb/spark/sql/connector/read/MongoInputPartitionHelper.java

```diff
@@ -78,6 +80,46 @@ static MongoInputPartition[] generateMongoBatchPartitions(
     }
   }
 
+  static MongoMicroBatchInputPartition[] generateMicroBatchPartitions(
+      final StructType schema,
+      final ReadConfig readConfig,
+      final BsonTimestampOffset start,
+      final BsonTimestampOffset end) {
+
+    List<BsonDocument> partitionPipeline = generatePipeline(schema, readConfig);
+
+    BsonTimestamp startTimestamp = start.getBsonTimestamp();
+    BsonTimestamp endTimestamp = end.getBsonTimestamp();
+
+    List<BsonTimestamp> partitions = new ArrayList<>();
+    partitions.add(startTimestamp);
+
+    if (startTimestamp.getTime() >= 0) {
+      int partitionCount = readConfig.getMicroBatchMaxPartitionCount();
+      int totalSecondsDiff = endTimestamp.getTime() - startTimestamp.getTime();
+      int numberOfBatches = Math.min(totalSecondsDiff, partitionCount);
+      int incPerBatch = (int) Math.ceil((double) totalSecondsDiff / numberOfBatches);
+
+      BsonTimestamp previous = startTimestamp;
+      while (previous.getTime() + incPerBatch < endTimestamp.getTime()) {
+        BsonTimestamp next = new BsonTimestamp(previous.getTime() + incPerBatch, 0);
+        partitions.add(next);
+        previous = next;
+      }
+    }
+    partitions.add(endTimestamp);
+
+    return IntStream.range(1, partitions.size())
+        .mapToObj(
+            i ->
+                new MongoMicroBatchInputPartition(
+                    i,
```

Comment:
1. Previously `MongoMicroBatchStream.planInputPartitions` used 0 as the minimal `MongoInputPartition.partitionId`, here 1 is used. Is that important?
2. Previously `MongoMicroBatchStream.planInputPartitions` remembered all used `partitionId` values (by remembering the largest one) and avoided reusing the same `partitionId` if called multiple times (its documentation states that _"will be called multiple times"_). With the changes in this PR, `partitionId` of partitions produced by multiple calls to `MongoMicroBatchStream.planInputPartitions` will intersect. Is this OK? If yes, what is the purpose of `MongoInputPartition.partitionId`?

Reply:
Partition Id is only used for logging purposes and is used to determine if a cursor has been opened & subsequently closed.

I considered reusing an ever increasing Id - but as these partitions are batch based and all partitions have to be consumed together within that batch before the next batch is created, I didn't consider it necessary.

So in short, the purpose of the Id is limited but occasionally useful in debugging.